pkg/config/msp_test.go

```diff
@@ -94,6 +108,799 @@ func TestMSPToProtoFailure(t *testing.T) {
+
+	gt.Expect(config).To(Equal(expectedConfigProto))
+}
+
+func TestAddRootCAToMSPFailure(t *testing.T) {
+	t.Parallel()
+	gt := NewGomegaWithT(t)
+
+	channelGroup, err := baseApplicationChannelGroup()
+	gt.Expect(err).ToNot(HaveOccurred())
+	config := &cb.Config{
+		ChannelGroup: channelGroup,
+	}
+
+	tests := []struct {
+		spec        string
+		cert        *x509.Certificate
+		expectedErr string
+	}{
+		{
+			spec: "invalid key usage",
+			cert: &x509.Certificate{
+				KeyUsage: x509.KeyUsageKeyAgreement,
+			},
+			expectedErr: "certificate KeyUsage must be x509.KeyUsageCertSign",
+		},
+		{
+			spec: "certificate is not a CA",
+			cert: &x509.Certificate{
+				IsCA:     false,
+				KeyUsage: x509.KeyUsageCertSign,
+			},
+			expectedErr: "certificate must be a CA certificate",
+		},
+	}
+
+	for _, tc := range tests {
+		tc := tc
+		t.Run(tc.spec, func(t *testing.T) {
+			gt := NewGomegaWithT(t)
+			err = AddRootCAToMSP(tc.cert, "Org1", config)
+			gt.Expect(err).To(MatchError(tc.expectedErr))
+		})
+	}
+}
+
+// creates a CRL with an empty list of revoked certificates
+// and returns it as ASN.1 encoded bytes
+func createCRLBytesForTest() []byte {
+	privKey := &ecdsa.PrivateKey{
+		PublicKey: ecdsa.PublicKey{
+			Curve: elliptic.P256(),
+			X:     big.NewInt(7),
+			Y:     big.NewInt(7),
+		},
+		D: big.NewInt(13),
+	}
+
+	caCert := &x509.Certificate{
+		KeyUsage:  x509.KeyUsageKeyEncipherment | x509.KeyUsageDigitalSignature | x509.KeyUsageCertSign,
+		IsCA:      true,
+		PublicKey: privKey.PublicKey,
+	}
+
+	signatureAlgorithm := pkix.AlgorithmIdentifier{
+		// copied this from ecdsa.oidSignatureECDSAWithSHA256
+		Algorithm: asn1.ObjectIdentifier{1, 2, 840, 10045, 4, 3, 2},
+	}
+	tbsCertList := pkix.TBSCertificateList{
+		Version:             1,
+		Signature:           signatureAlgorithm,
+		Issuer:              caCert.Subject.ToRDNSequence(),
+		RevokedCertificates: []pkix.RevokedCertificate{},
+	}
+	certificateList := pkix.CertificateList{
+		TBSCertList:        tbsCertList,
+		SignatureAlgorithm: signatureAlgorithm,
+	}
+
+	crlBytes, err := asn1.Marshal(certificateList)
+	if err != nil {
+		log.Fatalf("creating CRL failed: %v", err)
+	}
+
+	return crlBytes
+}
+
+func TestRevokeCertificateFromMSP(t *testing.T) {
+	t.Parallel()
+	gt := NewGomegaWithT(t)
+
+	channelGroup, err := baseApplicationChannelGroup()
+	gt.Expect(err).ToNot(HaveOccurred())
+	config := &cb.Config{
+		ChannelGroup: channelGroup,
+	}
+
+	crlBytes := createCRLBytesForTest()
+
+	// update Org1 with the CRL
+	org, err := getApplicationOrg(config, "Org1")
+	gt.Expect(err).NotTo(HaveOccurred())
+	fabricMSPConfig, err := getFabricMSPConfig(org)
+	gt.Expect(err).NotTo(HaveOccurred())
+	fabricMSPConfig.RevocationList = append(fabricMSPConfig.RevocationList, crlBytes)
+	err = addMSPConfigToOrg(org, fabricMSPConfig)
+	gt.Expect(err).NotTo(HaveOccurred())
+	config.ChannelGroup.Groups[ApplicationGroupKey].Groups["Org1"] = org
+
+	cert := &x509.Certificate{
```

Comment:
Please see earlier comment - I really think these tests need to use real certs and keys.

Reply:
Done